commcid.go

```diff
@@ -76,12 +82,125 @@ func CIDToCommitment(c cid.Cid) (FilMultiCodec, FilMultiHash, []byte, error) {
+// the assumption that the rest of the tree will be padded out (e.g. with zeros)
+func Fr32PaddedSizeToV1TreeHeight(size uint64) uint8 {
+	if size <= 32 {
+		return 0
+	}
+
+	// Calculate the floor of log2(size)
+	b := 63 - bits.LeadingZeros64(size)
+	// Leaf size is 32 == 2^5
+	b -= 5
+
+	// Check if the size is a power of 2 and if not then add one since the tree will need to be padded out
+	if 32<<b < size {
+		b += 1
+	}
+	return uint8(b)
+}
+
+// UnpaddedSizeToV1TreeHeight calculates the height of the piece tree given the data that's meant to be encoded in the
+// tree before any FR32 padding is applied. Because pieces are only defined on binary trees of FR32 encoded data if the
+// size is not a power of 2 after the FR32 padding is applied it will be rounded up to the next one under the assumption
+// that the rest of the tree will be padded out (e.g. with zeros)
+func UnpaddedSizeToV1TreeHeight(size uint64) (uint8, error) {
+	if size*128 < size {
+		return 0, fmt.Errorf("unsupported size: too big")
+	}
+
+	paddedSize := size * 128 / 127
+	if paddedSize*127 != size*128 {
+		paddedSize += 1
+	}
+
+	return Fr32PaddedSizeToV1TreeHeight(paddedSize), nil
+}
+
+// UnpaddedSizeToV1TreeHeightAndPadding calculates the height of the piece tree given the data that's meant to be
+// encoded in the tree before any FR32 padding is applied. Because pieces are only defined on binary trees of FR32
+// encoded data if the size is not a power of 2 after the FR32 padding is applied it will be rounded up to the next one
+// under the assumption that the rest of the tree will be padded out (e.g. with zeros). The amount of data padding that
+// is needed to be applied is returned alongside the tree height.
+func UnpaddedSizeToV1TreeHeightAndPadding(dataSize uint64) (uint8, uint64, error) {
+	if dataSize*128 < dataSize {
+		return 0, 0, fmt.Errorf("unsupported size: too big")
+	}
+
+	if dataSize < 127 {
+		return 0, 0, fmt.Errorf("unsupported size: too small")
+	}
+
+	fr32DataSize := dataSize * 128 / 127
+	// If the FR32 padding doesn't fill an exact number of bytes add up to 1 more byte of zeros to round it out
+	if fr32DataSize*127 != dataSize*128 {
+		fr32DataSize += 1
+	}
+
+	treeHeight := Fr32PaddedSizeToV1TreeHeight(fr32DataSize)
+	paddedFr32DataSize := uint64(32) << treeHeight
+	paddedDataSize := paddedFr32DataSize / 128 * 127
+	padding := paddedDataSize - dataSize
+
+	return treeHeight, padding, nil
+}
+
+// DataCommitmentV1ToPieceMhCID converts a raw data commitment and the height of the commitment tree
+// (i.e. log_2(padded data size in bytes) - 5, because 2^5 is 32 bytes which is the leaf node size) to a CID
+// by adding:
+// - codec: cid.Raw
+// - hash type: multihash.SHA2_256_TRUNC254_PADDED_BINARY_TREE
+//
+// The helpers UnpaddedSizeToV1TreeHeight and Fr32PaddedSizeToV1TreeHeight may help in computing tree height
+func DataCommitmentV1ToPieceMhCID(commD []byte, unpaddedDataSize uint64) (cid.Cid, error) {
```

Comment:
Allow one to supply `0` as "don't care" ?
```suggestion
func DataCommitmentV1ToPieceMhCID(commD []byte, optionalPayloadSize uint64) (cid.Cid, error) {
```